conn.go

```diff
@@ -58,114 +68,119 @@ func NewTimeoutConn(conn net.Conn, timeout time.Duration) (*Conn, error) {
-// Bytes written are prefixed with 32-bit header specifying the total size
-// of the data unit (message + 4 byte header), in network (big-endian) order.
-// http://www.ietf.org/rfc/rfc4934.txt
-func (c *Conn) writeDataUnit(x []byte) error {
-	logXML("<-- WRITE DATA UNIT -->", x)
-	s := uint32(4 + len(x))
+// writeRequest writes a single EPP request (x) for writing on c.
+// writeRequest can be called from multiple goroutines.
+func (c *Conn) writeRequest(x []byte) error {
+	c.mWrite.Lock()
+	defer c.mWrite.Unlock()
 	if c.Timeout > 0 {
 		c.Conn.SetWriteDeadline(time.Now().Add(c.Timeout))
 	}
-	err := binary.Write(c.Conn, binary.BigEndian, s)
-	if err != nil {
-		return err
-	}
-	_, err = c.Conn.Write(x)
-	return err
+	return writeDataUnit(c.Conn, x)
 }
 
-// readResponse reads a single EPP response from c and parses the XML into req.
+// readResponse dequeues and returns a EPP response from c.
 // It returns an error if the EPP response contains an error Result.
-func (c *Conn) readResponse(res *Response) error {
-	err := c.readDataUnit()
-	if err != nil {
-		return err
-	}
-	err = IgnoreEOF(scanResponse.Scan(c.decoder, res))
-	if err != nil {
-		return err
-	}
-	if res.Result.IsError() {
-		return &res.Result
+// readResponse can be called from multiple goroutines.
+func (c *Conn) readResponse() (*Response, error) {
+	select {
+	case res := <-c.responses:
+		if res == nil {
+			return res, c.readErr
+		}
+		if res.Result.IsError() {
+			return nil, &res.Result
+		}
+		return res, nil
+	case <-c.done:
+		return nil, net.ErrClosed
 	}
-	return nil
 }
 
-// readDataUnit reads a single EPP message from c into
-// c.buf. The bytes in c.buf are valid until the next
-// call to readDataUnit.
-func (c *Conn) readDataUnit() error {
-	c.reset()
-	var s int32
-	if c.Timeout > 0 {
-		c.Conn.SetReadDeadline(time.Now().Add(c.Timeout))
-	}
-	err := binary.Read(c.Conn, binary.BigEndian, &s)
-	if err != nil {
-		return err
-	}
-	s -= 4 // https://tools.ietf.org/html/rfc5734#section-4
-	if s < 0 {
-		return io.ErrUnexpectedEOF
+func (c *Conn) readLoop() {
+	defer close(c.responses)
+	timeout := c.Timeout
+	r := &io.LimitedReader{R: c.Conn}
+	decoder := xml.NewDecoder(r)
+	for {
+		if timeout > 0 {
+			c.Conn.SetReadDeadline(time.Now().Add(timeout))
+		}
+		n, err := parseDataUnit(c.Conn)
+		if err != nil {
+			c.readErr = err
+			return
+		}
+		r.N = int64(n)
+		res := &Response{}
+		err = IgnoreEOF(scanResponse.Scan(decoder, res))
+		if err != nil {
+			c.readErr = err
+			return
+		}
+		c.responses <- res
 	}
-	lr := io.LimitedReader{R: c.Conn, N: int64(s)}
-	n, err := c.buf.ReadFrom(&lr)
+}
+
+// writeDataUnit writes x to w.
+// Bytes written are prefixed with 32-bit header specifying the total size
+// of the data unit (message + 4 byte header), in network (big-endian) order.
+// http://www.ietf.org/rfc/rfc4934.txt
+func writeDataUnit(w io.Writer, x []byte) error {
+	logXML("<-- WRITE DATA UNIT -->", x)
+	s := uint32(4 + len(x))
```

Comment:
Why `uint32` here and `int32` in `parseDataUnit`?

Reply:
The spec is silent on whether it’s signed or unsigned. Hah. Making it uint32 everywhere.